### ISSUE TITLE
"total_cost" are always "nan"

### ISSUE THREAD

**Issue by NUAAXQ** [qq456cvb/VoteNet#9]
[0816 09:26:13 @monitor.py:474] param-summary/voting1/W-rms: 0.31994
[0816 09:26:13 @monitor.py:474] param-summary/voting2/W-rms: 0.23016
[0816 09:26:13 @monitor.py:474] sem_accuracy: nan
[0816 09:26:13 @monitor.py:474] sem_accuracy/SMA: nan
[0816 09:26:13 @monitor.py:474] total_cost: nan
[0816 09:26:13 @monitor.py:474] total_cost/SMA: nan
[0816 09:26:13 @base.py:275] Start Epoch 39 ...

When I train the model, I find that the total_cost are always nan. Is there something wrong with the loss computation? Did anyone come across the same problem?

**Comment by qq456cvb** [comment]
It is because once vote points are computed and grouped, there is a chance that no "positive" vote samples exists, in which case the loss computed by tf.gather_nd would always be "nan" (and the first dimension would be zero after tf.gather_nd). Since the total cost is a moving average, any single "nan" cost will result the final cost to be "nan". Anyway, it is better to check if the cost is "nan" and set it to tf.constant(zero) to prevent the total cost to be "nan", which may invalidate the gradient backprop.

**Comment by NUAAXQ** [comment]
> It is because once vote points are computed and grouped, there is a chance that no "positive" vote samples exists, in which case the loss computed by tf.gather_nd would always be "nan" (and the first dimension would be zero after tf.gather_nd). Since the total cost is a moving average, any single "nan" cost will result the final cost to be "nan". Anyway, it is better to check if the cost is "nan" and set it to tf.constant(zero) to prevent the total cost to be "nan", which may invalidate the gradient backprop.

**Comment by NUAAXQ** [comment]
Add the following code after line 84 in the model.py
        if  np.isnan(vote_reg_loss):
            vote_reg_loss = tf.constant(zero)

Do you mean that? By the way, did you also meet the same problem in your training?

**Comment by qq456cvb** [comment]
Yes, I also met this problem. By the way, the correct way to do this is something like `vote_reg_loss = tf.cond(tf.isnan(vote_reg_loss), tf.constant(zero), vote_reg_loss)`. We need to define the operation on the graph.

**Comment by qq456cvb** [comment]
Right. This happens when a batch has no positive vote/proposal samples: the `gather_nd` results are empty, `reduce_mean` over an empty tensor is NaN, and since `total_cost` is a moving average a single NaN poisons it from then on.

The fix is to guard the reductions so an empty batch contributes 0 instead of NaN. Something like:

```python
def safe_mean(x):
    return tf.cond(tf.size(x) > 0,
                   lambda: tf.reduce_mean(x),
                   lambda: tf.constant(0.0))
```

Then use `safe_mean(...)` for the losses computed over `positive_idxes` (objectness, center, heading, size, semantic). That keeps the gradients finite even on batches with no positives. Leaving this open as the reference for the issue.
